File: funcs_util/funcs.py
import pandas as pd
import numpy as np

def baseline_hazard(df_input):
    # baseline_hazard function with Efron correction
    # df_input: n_sample*(time0,time1,event,risk_score)
    # output: n_event_time*(time,h)
    cols = ["time0","time1","event","risk_score"]
    df_input = pd.DataFrame(df_input, columns = cols)
    df_input[cols] = df_input[cols].apply(pd.to_numeric, errors='coerce', axis=1)

    event_time,ties_vector = np.unique(df_input['time1'][df_input['event'] == 1],return_counts=True)
    n_event_time = len(event_time)
    Max_tie = max(ties_vector)

    rs_matrix = np.tile(np.array([df_input['risk_score'].to_numpy()]).transpose(),(1,n_event_time))
    exprs_matrix = np.exp(rs_matrix)
    # n_sample * n_event_time(sorted)

    atrisk_matrix = ((df_input['time0'].to_numpy() < event_time[:,None]) & (df_input['time1'].to_numpy() >= event_time[:,None])).transpose()
    event_matrix = ((df_input['time1'].to_numpy()*df_input['event'].to_numpy()) == event_time[:,None]).transpose()
    # n_sample * n_event_time(sorted)

    # tie sum (for efron correction)
    tie_sum_vector = np.sum(exprs_matrix * atrisk_matrix,axis=0)
    # length: n_event_time(sorted)

    # tie mean (for efron correction)
    tie_mean_vector = np.sum(exprs_matrix * event_matrix,axis=0)/ties_vector

    # efron matrix
    efron_matrix = np.tile(np.array(range(Max_tie)),(n_event_time,1))
    # n_event_time*Max_tie

    # mask matrix to locate the ties
    mask_matrix = ((np.array(range(Max_tie))+1) <= ties_vector[:,None])

    h = np.tile(np.expand_dims(tie_sum_vector, axis=1),(1,Max_tie))-np.tile(np.expand_dims(tie_mean_vector, axis=1),(1,Max_tie))*efron_matrix
    h = 1/h
    h[np.isinf(h)] = 0
    h = h*mask_matrix
    h = np.sum(h,axis = 1)
    
    event_time = np.insert(event_time, 0, 0)
    h = np.insert(h, 0, 0)

    output = pd.DataFrame({'time': event_time, 'h': h})
    return output

def survprob(time_of_interest,haz,test_risk_score):
    # time_of_interest: a list of sorted time points on which the survival probability will be calculated. It is the time window after the last observation time (last_obs_time) of each subject in test_risk_score.
    # haz: n_times_train*(time,h0)
    # test_risk_score: n_obs*(id,last_obs_time,risk_score)
    time_of_interest = np.sort(np.array(time_of_interest))
    haz = pd.DataFrame(haz,columns = ['time','h'])
    test_risk_score = pd.DataFrame(test_risk_score,columns = ['id','last_obs_time','risk_score'])
    
    _,id_count = np.unique(test_risk_score["id"],return_counts=True)
    if sum(id_count>1)>0:
        raise ValueError('Each subject_id should only have one row. Please check test_risk_score')
    
    timeofinterest_matrix = np.add.outer(test_risk_score['last_obs_time'].to_numpy(), time_of_interest)
    all_times = np.unique(np.concatenate((test_risk_score['last_obs_time'].to_numpy(),
                                          timeofinterest_matrix.flatten(),
                                          haz['time'].to_numpy()), 
                                         axis=None))

    haz_alltimes = pd.merge(pd.DataFrame({'time': all_times}), haz, how="outer")
    haz_alltimes.fillna(0, inplace=True)

    survpred = np.outer(np.exp(test_risk_score['risk_score']),haz_alltimes['h'])
    survpred = np.cumsum(survpred,axis=1)
    survpred = np.exp(-survpred)

    index_timeofinterest = np.array([np.searchsorted(all_times, timeofinterest_matrix[i]) for i in range(len(timeofinterest_matrix))])
    index_last_obs_time = np.array([np.searchsorted(all_times, test_risk_score['last_obs_time'][i]) for i in range(len(timeofinterest_matrix))])

    output = np.array([(survpred[ii][index_timeofinterest[ii]])/(survpred[ii][index_last_obs_time[ii]]) for ii in range(len(survpred))])
    output = pd.DataFrame(output,columns = [str(k) for k in time_of_interest])
    output = pd.concat([test_risk_score[['id','last_obs_time']],output],axis = 1)
    return output

File: funcs_util/funcs.R
require(dplyr)
require(tidyr)
require(pec)
require(timeROC)

select <- dplyr::select
filter <- dplyr::filter

# baseline hazard function
baseline_hazard <- function(df_input){
  # baseline_hazard function with Efron correction
  # df_input: n_obs*(time0,time1,event,risk_score)
  # output: n_event_time*(time,h)
  df_input <- as.data.frame(as.matrix(df_input))
  names(df_input) <- c("time0","time1","event","risk_score")
  
  # unique event time
  event_time <- sort(unique(df_input$time1[(df_input$event == 1)]))
  n_event_time <- length(event_time)
  
  ties_vector <- unname(table(df_input$time1[(df_input$event == 1)]))
  Max_tie <- max(ties_vector)
  
  # predicted risk score at event time
  rs_matrix <- matrix(df_input$risk_score) %*% base::t(rep(1,n_event_time))
  exprs_matrix <- exp(rs_matrix)
  # n_obs * n_event_time(sorted)
  
  # index matrixs
  atrisk_matrix <- (outer(df_input$time0,event_time,"<") & outer(df_input$time1,event_time,">="))
  event_matrix <- outer(df_input$time1*df_input$event,event_time,"==")
  # n_obs * n_event_time(sorted)
  
  # tie sum (for efron correction)
  tie_sum_vector <- colSums(exprs_matrix * atrisk_matrix)
  # length: n_event_time(sorted)
  
  # tie mean (for efron correction)
  tie_mean_vector <- colSums(exprs_matrix * event_matrix)/ties_vector
  
  # efron matrix
  efron_matrix <- matrix(rep((seq(Max_tie)-1),n_event_time),nrow = n_event_time, byrow = T)
  # n_event_time*Max_tie
  
  # mask matrix to locate the ties
  mask_matrix <- outer(ties_vector,seq(Max_tie),">=")
  
  h <- outer(tie_sum_vector,rep(1,Max_tie),"*")-outer(tie_mean_vector,rep(1,Max_tie),"*")*efron_matrix
  h <- 1/h
  h[is.infinite(h)] <- 0
  h <- h*mask_matrix
  h <- rowSums(h)
  
  output <- data.frame(time = c(0,event_time), h = c(0,h))
  return(output)
}

#####
survprob <- function(time_of_interest=NULL,
                     haz=NULL,
                     test_risk_score=NULL){
  # time_of_interest: a list of sorted time points on which the survival probability will be calculated. 
  # It is the time window after the last observation time (last_obs_time) of each subject in test_risk_score.
  # haz: n_times_train*(time,h0)
  # test_risk_score: n_sample*(id,last_obs_time,risk_score)
  haz <- as.data.frame(haz)
  names(haz) <- c("time","h")
  
  test_risk_score <- as.data.frame(test_risk_score)
  names(test_risk_score) <- c("id","last_obs_time","rs")
  
  id_count <- test_risk_score$id %>% unique %>% table %>% max
  
  if(id_count>1){stop("Each subject_id should only have one row. Please check test_risk_score")}
  
  time_ofinterest_matrix <- outer(test_risk_score$last_obs_time,time_of_interest,"+")
  all_times <- c(as.vector(time_ofinterest_matrix),haz$time,test_risk_score$last_obs_time) %>%
    unique() %>%
    sort()
  
  haz_alltimes <- haz %>%
    complete(time = all_times,fill = list(h=0))
  
  h_tmp <- outer(exp(test_risk_score$rs),haz_alltimes$h)
  cumh_tmp <- t(apply(h_tmp, 1, cumsum)) %>% as.matrix
  
  survpred_output <- sapply(seq(nrow(test_risk_score)),function(j){
    index_timeofinterest_tmp <- (all_times %in% (time_ofinterest_matrix[j,] %>% unlist))
    index_time_tmp <- (all_times == test_risk_score$last_obs_time[j])
    output_tmp <- exp(-cumh_tmp[j, index_timeofinterest_tmp]+cumh_tmp[j,index_time_tmp])
    return(output_tmp)
  }) %>% base::t()
  
  survpred_output <- as.data.frame(survpred_output)
  names(survpred_output) <- paste0('t',time_of_interest)
  survpred_output <- cbind(test_risk_score,survpred_output)
  
  return(survpred_output)
}
